# Portada

### Estudiantes:

    Jason Solano 

    Percy herrara
### Curso
    Inteligencia de negocios y mineria de datos
### Trabajo
    Projecto Final 

# Fase de Entendimiento del negocio

## Objetivos del negocio

- Facilitar el pronóstico de cuentas que pueden ser incobrables a un largo plazo para poder establecer un arreglo de pago para realizar una liquidación adecuada de la cuenta.


- Poder ofrecer una herramienta para el diseño robusto de campañas de cobro de carteras de crédito, las cuales pueden estar conformadas con cuentas que pueden ser incobrables.


## Criterios de éxito en términos de negocio

- Lograr la identificación exitosa de cuentas que pueden ser incobrables para una entidad financiera, reduciendo la cantidad de cuentas incobrables al menos en 30% de acuerdo a los datos históricos de la entidad financiero.


- Mostrar a las entidades financieros cuales son los atributos de los clientes  que lo puede llevar a ser una cuenta incobrable para poder facilitar la estadística descriptiva de la entidad financiera.

## Objetivos de minería de datos

- Identificar por medio de modelos de predicción potenciales cuentas que puedan ser incobrables en un futuro, para ofrecer una liquidación por medio de un arreglo de pago.

- Identificar cuales son los atributos que más influyen para determinar si una cuenta se le debe realizar la oferta de liquidación debido a que  es catalogada como incobrabl


## Criterios de éxito para la minería de datos

- Obtener al menos un modelo que presente un accuracy (métrica de precisión para exactitud) mayor a un 70% con la utilización de datos externos a los utilizados en el entrenamiento.

- Optimizar con diferentes parámetros al menos tres modelos utilizados para la predicción en la categoría de clasificación binaria.


# Fase de Entendimiento de los datos

## Descripción de los datos

### Información del conjunto de datos


El conjunto de datos es sobre información de campaña de recoleccion de dineros de cuentas en mora, es la informacion sobre clientes contactados, con una oferta concreta para cerrar su cuenta en mora por un monto inferior al actual.

El conjunto de datos esta separado por dos archivos CSV, los cuales vamos a unir

In [1]:
import pandas as pd
import matplotlib.pyplot as plt

df1 = pd.read_csv('proyectoF1T.csv')
df2 = pd.read_csv('proyectoF2F.csv')
print(df1.shape)
df1.head()

(2595, 31)


id  approval_amount recommendation                   redeem_ts  \
0  503603788            250.0        APPROVE  2019-01-24 09:50:41.115000   
1  399999590             80.0        APPROVE  2019-01-24 17:34:47.768000   
2  330633304            956.0        APPROVE  2019-01-24 18:18:53.120000   
3  422247835              9.0        APPROVE  2019-01-24 18:36:06.694000   
4  427183074            250.0        APPROVE  2019-01-25 08:47:49.726000   

   contacted          email_contacted_ts            sms_contacted_ts  \
0      False                         NaN  2019-01-24 09:36:27.784000   
1      False  2019-01-24 12:16:54.626000                         NaN   
2      False  2019-01-24 13:17:29.359000  2019-01-24 13:15:59.087000   
3      False                         NaN  2019-01-24 13:19:17.988000   
4      False  2019-01-24 09:08:46.690000  2019-01-24 09:08:20.447000   

     birth_dt state         city         ...           account_payment_method  \
0  1988-05-21    CO       Aurora         ...                           MANUAL   
1  1998-01-13    TX  Weatherford         ...                           MANUAL   
2  1975-12-09    WA      Othello         ...                           MANUAL   
3  1985-09-09    CA     Imperial         ...                           MANUAL   
4  1980-11-23    OH    Cleveland         ...                           MANUAL   

  collections_category  days_past_due  number_of_payments amount_financed  \
0                  NaN              0                  52             NaN   
1                  NaN              0                  26             0.0   
2                  NaN              0                  12             0.0   
3                  NaN              0                  26             0.0   
4                  NaN              0                  26             0.0   

   data_correction merchant_id    source initial_decision  last_payment_amount  
0            False    14206566  MERCHANT         APPROVED                70.34  
1            False   107313386  INTERNET         APPROVED                  NaN  
2            False   230266536  MERCHANT         APPROVED                  NaN  
3            False   332752404  MERCHANT         APPROVED                  NaN  
4            False   358414568  MERCHANT         APPROVED                  NaN  

[5 rows x 31 columns]

### Se realiza la unión de los dos datasets

Por lo cual se puede ver la cantidad de registros y columnas.

Cantidad Registros: 6095

Cantidad de columnas: 31

In [2]:
df1 = df1.append(df2)
df2 = None    
print(df1.shape)

(6095, 31)


### Se observan las variables del dataset

In [3]:
df1.columns

Index(['id', 'approval_amount', 'recommendation', 'redeem_ts', 'contacted',
       'email_contacted_ts', 'sms_contacted_ts', 'birth_dt', 'state', 'city',
       'zip_code', 'rent_or_own', 'months_at_current_residence',
       'years_at_current_residence', 'direct_deposit', 'black_listed',
       'language_preference', 'military_applicant', 'payment_frequency',
       'scheduled_payment_amt', 'past_due_amt', 'account_payment_method',
       'collections_category', 'days_past_due', 'number_of_payments',
       'amount_financed', 'data_correction', 'merchant_id', 'source',
       'initial_decision', 'last_payment_amount'],
      dtype='object')

## Explicación de las columnas

- id
    - Id de cada fila.
- merchant_id
    - Id para cada comerciante.
- last_payment_amount
    - La cantidad del último pago realizado.
- birth_dt
    - Fecha de cumpleaños de cada cliente
- redeem_ts
    - Personas a las cuales se les debe ofrecer la oferta de liquidación, fecha que se hizo.
- account_payment_method
    - Forma de pago configurada para la cuenta.
- approval_amount
    - Monto aprobado del préstamo.
- scheduled_payment_amt
    - Monto de la cuota configurada para el préstamo.
- days_past_due
    - Días en que acumula en mora la cuenta.
- past_due_amt
    - Monto que se encuentra en atraso para la cuenta.
- zip_code
    - Código postal del cliente al cual le pertenece la cuenta.
- age
    - Edad del cliente de la cuenta.
- city
    - Ciudad de domicilio del cliente.
- sms_contacted_ts
    - Fecha en que se le contactó por medio de mensaje de texto al celular.
- state
    - Estado de domicilio de la persona de la cuenta.
- years_at_current_residence
    - Años viviendo en la actual residencia.
- months_at_current_residence
    - Meses viviendo en la actual residencia.
- source
    - Fuente desde donde se genero el préstamo (Internet, Agencia, etc).
- collections_category
    - Columna no especificamente descripta.
- email_contacted_ts
    - Fecha en que se le contactó por medio de correo electrónico.
- payment_frequency
    - Frecuencia de pago de la cuenta.
- number_of_payments
    - Cantidad de pagos necesarios para cancelar la cuenta.
- rent_or_own
    - Es propietario de vivienda o renta.
- initial_decision
    - Decisión inicial al crear la cuenta (Aprobación o Denegar).
- language_preference
    - Lenguaje en que prefiere el cliente ser contactado (Español o Inglés).
- contacted
    - Si se logró contactar anteriormente.
- data_correction
    - Si se tuvo que corregir los datos del cliente en algún momento (True, False).
- black_listed
    - Si en algún momento el cliente fue incluido en lista negra por fraude.
- direct_deposit
    - Si la el préstamo está configurado para recibir los pagos de forma automática por depósito.
- recommendation
    - Recomendación que se señaló al formalizar la cuenta (Aprobar o Denegar).
- military_applicant
    - Si el cliente es veterano o actualmente miembro de la fuerza militar.


### Tipos de datos

    id                               int64

    approval_amount                float64

    recommendation                  string

    redeem_ts                         date

    contacted                         bool

    email_contacted_ts                date

    sms_contacted_ts                  date

    birth_dt                          date

    state                           string

    city                            string

    zip_code                         int64

    rent_or_own                     string

    months_at_current_residence      int64

    years_at_current_residence       int64

    direct_deposit                    bool

    black_listed                      bool

    language_preference             string

    military_applicant                bool

    payment_frequency               string

    scheduled_payment_amt          float64

    past_due_amt                   float64

    account_payment_method          string

    collections_category            string

    days_past_due                    int64

    number_of_payments               int64

    amount_financed                float64

    data_correction                   bool

    merchant_id                      int64

    source                          string

    initial_decision                string

    last_payment_amount            float64

### Datos generales del conjunto de datos

In [4]:
df1.describe()

id  approval_amount       zip_code  \
count  6.095000e+03      6095.000000    6095.000000   
mean   4.565517e+08       557.357670   61592.610008   
std    1.730670e+08       371.192573   28386.361369   
min    4.396600e+04         3.000000       0.000000   
25%    3.692606e+08       274.000000   33012.000000   
50%    4.760580e+08       487.000000   76119.000000   
75%    5.770807e+08       764.000000   85041.000000   
max    8.277996e+08      2899.000000  320568.000000   

       months_at_current_residence  years_at_current_residence  \
count                  6095.000000                 6095.000000   
mean                      2.070057                    3.547990   
std                       3.213727                    3.137995   
min                       0.000000                    0.000000   
25%                       0.000000                    1.000000   
50%                       0.000000                    3.000000   
75%                       3.000000                    6.000000   
max                      24.000000                   48.000000   

       scheduled_payment_amt  past_due_amt  days_past_due  number_of_payments  \
count            6095.000000   6095.000000    6095.000000         6095.000000   
mean              107.171736     94.349810      14.462510           32.452174   
std                89.448125    357.118752      46.452751           14.133418   
min               -83.120000      0.000000       0.000000           12.000000   
25%                48.040000      0.000000       0.000000           24.000000   
50%                80.430000      0.000000       0.000000           26.000000   
75%               134.450000      0.000000       0.000000           52.000000   
max               696.710000   3714.450000     337.000000           52.000000   

       amount_financed   merchant_id  last_payment_amount  
count           2855.0  6.095000e+03          3044.000000  
mean               0.0  1.991524e+08           117.082050  
std                0.0  1.718332e+08           102.038427  
min                0.0  2.110000e+02         -1344.000000  
25%                0.0  1.853463e+07            62.007500  
50%                0.0  1.660589e+08            93.760000  
75%                0.0  3.524943e+08           146.012500  
max                0.0  4.902396e+08          2240.740000

Por medio del describe se nos muestra un summary de cuartiles, cantidad de registros,la media y desviación estandar

## Exploración de los datos y verificación de datos

Para esta sección se aplicaran ambos casos, aunque también se ha aplicado exploración de los datos en la sección anterior.

### Analisis de variables nulas

En la siguiente sección se evaluara la cantidad de nulos, esta cantidad sera mostrada por medio del porcentaje de nulos para cada columna.

In [5]:
df1.isna().mean().round(4)

id                             0.0000
approval_amount                0.0000
recommendation                 0.0000
redeem_ts                      0.5742
contacted                      0.0000
email_contacted_ts             0.0786
sms_contacted_ts               0.3993
birth_dt                       0.0000
state                          0.0000
city                           0.0003
zip_code                       0.0000
rent_or_own                    0.0000
months_at_current_residence    0.0000
years_at_current_residence     0.0000
direct_deposit                 0.0028
black_listed                   0.0000
language_preference            0.0000
military_applicant             0.0000
payment_frequency              0.0000
scheduled_payment_amt          0.0000
past_due_amt                   0.0000
account_payment_method         0.0000
collections_category           0.4167
days_past_due                  0.0000
number_of_payments             0.0000
amount_financed                0.5316
data_correct

Se puede observar existe varias columnas que contienen un alto porcentaje de nulos, por lo cual en la siguiente fase seran evaluadas.

### Análisis para conversión de datos

Algunas variables presentan valores nulos, que en una representacion binario serian de un valor cero y si contienen datos serian un valor 1.

**sms_contacted_ts**

Esta columna aplica dado que los nulos equivalen a personas que no fueron contactadas por SMS, esta información si es importante para nuestro análisis, en la siguiente fase remplazaremos estos valores por valores booleanos para identificar las personas contactadas por este medio.

**emal_contacted**

Esta columna aplica dado que los nulos equivalen a personas que no fueron contactadas por email, esta información si es importante para nuestro análisis, en la siguiente fase remplazaremos estos valores por valores booleanos para identificar las personas contactadas por este medio.

**redeem_ts**

Un caso muy similar a las otra columnas donde los valores nulos serian personas que no se le presentaron oferta, y las instancias con valores de fecha serian las personas que si se le presentaron una oferta.

**direct_deposit**

Columna Booleane que indica si se hizo un depósito directo, para el caso de los valores nulos que se encuentran seran tomados como falsos.

# Fase de Preparación de los datos

## Limpieza de los datos y selección de datos

Durante esta sección veremos algunos datos descartados y limpieza de valores nulos.

### Realizamos la conversión de datos nulos
Como se menciona anteriormente tratamos las variables con valores nulos que pueden ser interpretados como booleanos.

In [6]:
df1.loc[~df1.email_contacted_ts.isnull(),"email_contacted_ts"]=True 
df1.loc[~df1.sms_contacted_ts.isnull(),"sms_contacted_ts"]=True 
df1.loc[~df1.direct_deposit.isnull(),"direct_deposit"]=True 
df1.loc[~df1.redeem_ts.isnull(),"redeem_ts"]=True 
df1["email_contacted_ts"].fillna(False, inplace = True) 
df1["sms_contacted_ts"].fillna(False, inplace = True)
df1["direct_deposit"].fillna(False, inplace = True)
df1["redeem_ts"].fillna(False, inplace = True) 
df1.head()

id  approval_amount recommendation  redeem_ts  contacted  \
0  503603788            250.0        APPROVE       True      False   
1  399999590             80.0        APPROVE       True      False   
2  330633304            956.0        APPROVE       True      False   
3  422247835              9.0        APPROVE       True      False   
4  427183074            250.0        APPROVE       True      False   

   email_contacted_ts  sms_contacted_ts    birth_dt state         city  \
0               False              True  1988-05-21    CO       Aurora   
1                True             False  1998-01-13    TX  Weatherford   
2                True              True  1975-12-09    WA      Othello   
3               False              True  1985-09-09    CA     Imperial   
4                True              True  1980-11-23    OH    Cleveland   

          ...           account_payment_method collections_category  \
0         ...                           MANUAL                  NaN   
1         ...                           MANUAL                  NaN   
2         ...                           MANUAL                  NaN   
3         ...                           MANUAL                  NaN   
4         ...                           MANUAL                  NaN   

   days_past_due  number_of_payments  amount_financed  data_correction  \
0              0                  52              NaN            False   
1              0                  26              0.0            False   
2              0                  12              0.0            False   
3              0                  26              0.0            False   
4              0                  26              0.0            False   

  merchant_id    source initial_decision  last_payment_amount  
0    14206566  MERCHANT         APPROVED                70.34  
1   107313386  INTERNET         APPROVED                  NaN  
2   230266536  MERCHANT         APPROVED                  NaN  
3   332752404  MERCHANT         APPROVED                  NaN  
4   358414568  MERCHANT         APPROVED                  NaN  

[5 rows x 31 columns]

Nuevamente se vuelve a comprobar los valores que son nulos.

In [7]:
df1.isna().mean().round(4)

id                             0.0000
approval_amount                0.0000
recommendation                 0.0000
redeem_ts                      0.0000
contacted                      0.0000
email_contacted_ts             0.0000
sms_contacted_ts               0.0000
birth_dt                       0.0000
state                          0.0000
city                           0.0003
zip_code                       0.0000
rent_or_own                    0.0000
months_at_current_residence    0.0000
years_at_current_residence     0.0000
direct_deposit                 0.0000
black_listed                   0.0000
language_preference            0.0000
military_applicant             0.0000
payment_frequency              0.0000
scheduled_payment_amt          0.0000
past_due_amt                   0.0000
account_payment_method         0.0000
collections_category           0.4167
days_past_due                  0.0000
number_of_payments             0.0000
amount_financed                0.5316
data_correct

### Eliminación de  columnas y filas que contengan datos nulos

En este proceso se realiza la primera escogencia de datos.

Al observar las siguientes columnas que contienen una gran cantidad de nulos seran borradas:

collections_category

amount_financed

last_payment_amount

Por otra parte se decide borrar un porcentaje pequeño de filas equivalente al 0.0003, 

pertenecientes a valores nulos



In [8]:
df1 = df1.drop(columns=['collections_category','amount_financed','last_payment_amount'])

In [9]:
df1.shape

(6095, 28)

Se borran el 0.0003 de filas pertenecientes a valores nulos de city

In [10]:
df1 = df1.dropna()
df1.shape

(6093, 28)

### Volvemos a observar el porcentaje de nulos

In [11]:
df1.isna().mean().round(4)

id                             0.0
approval_amount                0.0
recommendation                 0.0
redeem_ts                      0.0
contacted                      0.0
email_contacted_ts             0.0
sms_contacted_ts               0.0
birth_dt                       0.0
state                          0.0
city                           0.0
zip_code                       0.0
rent_or_own                    0.0
months_at_current_residence    0.0
years_at_current_residence     0.0
direct_deposit                 0.0
black_listed                   0.0
language_preference            0.0
military_applicant             0.0
payment_frequency              0.0
scheduled_payment_amt          0.0
past_due_amt                   0.0
account_payment_method         0.0
days_past_due                  0.0
number_of_payments             0.0
data_correction                0.0
merchant_id                    0.0
source                         0.0
initial_decision               0.0
dtype: float64

Por lo tanto en este momento el dataset queda sin valores nulos

### Se borran columnas que son que algún tipo de ID

Dichas  columnas no presentan valor para los algoritmos de clasificación, que contienen un ID unico para cada columna

In [12]:
df1 = df1.drop(columns=['merchant_id','id'])

In [13]:
df1.shape

(6093, 26)

Por lo tanto se han descartado u total de 5 variables y borrado todo los datos nulos

### Conclusión de preselección y limpieza de datos

Al terminar esta sección de limpieza de datos se identifican variables con alto porcentaje de datos nulos donde  es mejor eliminarlas ya que possen mas de un 40% de valores nulos.

Por otra parte existen columnas que son ID para cada fila, estas columnas tambien son borradas ya que no presenta un valor  para los algoritmos de machine learning.

## Transformaciones aplicadas a los datos

### Transformación de datos categoricos con un ordinal encoder

Para probar el random forest para detectar las variables mas importantes, tranfomaremos las variables categoricas de forma ordinal, ya que los desicion trees no se ven tan afectados por datos categoricos que son transformados por esta forma.

Los datos categoricos a transformar son:


redeem_ts

contacted

email_contacted_ts

sms_contacted_ts

state

city

payment_frequency

rent_or_own

account_payment_method

data_correction

source

initial_decision

direct_deposit

language_preference

recomendation

military_applicant

black_listed


In [14]:
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OrdinalEncoder

encoder = OrdinalEncoder()
df1.redeem_ts = encoder.fit_transform(df1.redeem_ts.values.reshape(-1, 1))
df1.contacted = encoder.fit_transform(df1.contacted.values.reshape(-1, 1))
df1.email_contacted_ts = encoder.fit_transform(df1.email_contacted_ts.values.reshape(-1, 1))
df1.sms_contacted_ts = encoder.fit_transform(df1.sms_contacted_ts.values.reshape(-1, 1))
df1.state = encoder.fit_transform(df1.state.values.reshape(-1, 1))
df1.city = encoder.fit_transform(df1.city.values.reshape(-1, 1))
df1.payment_frequency = encoder.fit_transform(df1.payment_frequency.values.reshape(-1, 1))
df1.rent_or_own = encoder.fit_transform(df1.rent_or_own.values.reshape(-1, 1))
df1.account_payment_method = encoder.fit_transform(df1.account_payment_method.values.reshape(-1, 1))
df1.data_correction = encoder.fit_transform(df1.data_correction.values.reshape(-1, 1))
df1.source = encoder.fit_transform(df1.source.values.reshape(-1, 1))
df1.initial_decision = encoder.fit_transform(df1.initial_decision.values.reshape(-1, 1))
df1.direct_deposit = encoder.fit_transform(df1.direct_deposit.values.reshape(-1, 1))
df1.language_preference = encoder.fit_transform(df1.language_preference.values.reshape(-1, 1))
df1.recommendation = encoder.fit_transform(df1.recommendation.values.reshape(-1, 1))
df1.military_applicant = encoder.fit_transform(df1.military_applicant.values.reshape(-1, 1))
df1.black_listed = encoder.fit_transform(df1.black_listed.values.reshape(-1, 1))

In [15]:
df1.head()

approval_amount  recommendation  redeem_ts  contacted  email_contacted_ts  \
0            250.0             0.0        1.0        0.0                 0.0   
1             80.0             0.0        1.0        0.0                 1.0   
2            956.0             0.0        1.0        0.0                 1.0   
3              9.0             0.0        1.0        0.0                 0.0   
4            250.0             0.0        1.0        0.0                 1.0   

   sms_contacted_ts    birth_dt  state    city  zip_code        ...         \
0               1.0  1988-05-21    5.0    73.0     80012        ...          
1               0.0  1998-01-13   41.0  1590.0     76086        ...          
2               1.0  1975-12-09   45.0  1086.0     99344        ...          
3               1.0  1985-09-09    4.0   684.0     92251        ...          
4               1.0  1980-11-23   33.0   264.0     44105        ...          

   military_applicant  payment_frequency  scheduled_payment_amt  past_due_amt  \
0                 0.0                4.0                  18.84           0.0   
1                 0.0                0.0                  32.88           0.0   
2                 0.0                1.0                 423.50           0.0   
3                 0.0                0.0                  18.15           0.0   
4                 0.0                0.0                  87.73           0.0   

   account_payment_method  days_past_due  number_of_payments  data_correction  \
0                     2.0              0                  52              0.0   
1                     2.0              0                  26              0.0   
2                     2.0              0                  12              0.0   
3                     2.0              0                  26              0.0   
4                     2.0              0                  26              0.0   

   source  initial_decision  
0     2.0               0.0  
1     0.0               0.0  
2     2.0               0.0  
3     2.0               0.0  
4     2.0               0.0  

[5 rows x 26 columns]

## Construcción de nuevos datos (atributos)

Para este caso vamos a construir una nueva variable age, apartir de la variable birth_dt, por lo tanto seria un nuevo atributo.

In [17]:
now = pd.Timestamp('now')
df1['birth_dt'] = pd.to_datetime(df1['birth_dt'])   
df1['age'] = (now - df1['birth_dt']).astype('<m8[Y]') 
df1 = df1.drop(columns=['birth_dt'])

In [18]:
df1.age.head()

0    31.0
1    22.0
2    44.0
3    34.0
4    39.0
Name: age, dtype: float64

## Selección de los datos

Para la selección final de datos vamos a aplicar un random forest el cual no permite escoger por medio de un nivel de importancia los atributos que mas no favorecen a la hora entrenar los algoritmos de predicción

### Balanceo de datos

En nuestro caso la columna **redeem_ts** viene siendo nuestro label para aplicar la clasificación, por lo tanto se realiza un balance entre cuendo se realizo oferta y no se realizo oferta.

Se convierte la columna **redeem_ts** en one hot coding

In [21]:
redeem = df1[(df1['redeem_ts'] == 1)]
noRedeem = df1[(df1['redeem_ts'] == 0)]

In [22]:
print(redeem.shape,noRedeem.shape)

(2595, 26) (3498, 26)


Al observar los datos se puede ver que existen más datos de no oferta a oferta, por lo cual se escogen 2595 filas de no oferta para realizar un correcto balance

In [23]:
dfBalanceado = pd.concat([redeem,noRedeem.sample(redeem.shape[0])])

### Probamos la extracción de los variables mas importantes utilizando el random forest

 Se separan los features y labels

In [24]:
labels = df1['redeem_ts']
features = df1.drop(columns=['redeem_ts'])

Se separan los datos de entrenamiento para el random forest

In [26]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(features,labels,test_size=0.15,random_state=13)

Se presenta una configuración recomenda para dicha tarea, mas adelante en la fase de modelado se aplicara una mejor escogencia de parámetros para este mismo algoritmo.

In [27]:
from sklearn.ensemble import RandomForestClassifier
RandomForestModel = RandomForestClassifier(min_samples_leaf=3,min_samples_split=20,n_estimators=500, max_depth= None, random_state=10)
RandomForestModel.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=3, min_samples_split=20,
                       min_weight_fraction_leaf=0.0, n_estimators=500,
                       n_jobs=None, oob_score=False, random_state=10, verbose=0,
                       warm_start=False)

#### Realizamos el análisis para obtener las variables mas importantes

In [28]:
importances = pd.DataFrame({'feature' :features.columns, 'importance':RandomForestModel.feature_importances_})

In [29]:
importances = importances.sort_values('importance', ascending = False).set_index('feature')
importances

importance
feature                                
account_payment_method         0.224321
approval_amount                0.188819
scheduled_payment_amt          0.083982
days_past_due                  0.064521
zip_code                       0.058695
age                            0.056809
past_due_amt                   0.055592
city                           0.055350
sms_contacted_ts               0.038639
state                          0.033293
years_at_current_residence     0.031959
months_at_current_residence    0.023558
source                         0.019083
email_contacted_ts             0.016261
payment_frequency              0.015773
number_of_payments             0.014557
rent_or_own                    0.007484
initial_decision               0.005961
language_preference            0.004128
contacted                      0.000663
data_correction                0.000315
black_listed                   0.000161
direct_deposit                 0.000074
military_applicant             0.000000
recommendation                 0.000000

Por lo tanto de acuerdo al análisis del random forest se puede observar que los datos mas impotantes que presentan almenos un 1% del total de importancia: 

In [30]:
importances.index[:16]

Index(['account_payment_method', 'approval_amount', 'scheduled_payment_amt',
       'days_past_due', 'zip_code', 'age', 'past_due_amt', 'city',
       'sms_contacted_ts', 'state', 'years_at_current_residence',
       'months_at_current_residence', 'source', 'email_contacted_ts',
       'payment_frequency', 'number_of_payments'],
      dtype='object', name='feature')